Rate limiting /1 (#26756)
dotnet/AspNetCore.Docs: aspnetcore/performance/rate-limit.md

---
title: Rate limiting middleware in ASP.NET Core
author: rick-anderson
description: Learn how limit requests in ASP.NET Core apps
ms.author: riande
ms.custom: mvc
ms.date: 8/28/2022
uid: performance/rate-limit
---

# Rate limiting middleware in ASP.NET Core

By [Arvin Kahbazi](https://github.com/Kahbazi) and [Rick Anderson](https://twitter.com/RickAndMSFT)

:::moniker range=">= aspnetcore-7.0"

> The [Microsoft.AspNetCore.RateLimiting](https://www.nuget.org/packages/Microsoft.AspNetCore.RateLimiting) NuGet package provides rate limiting middleware. Apps configure rate limiting policies and then attach the policies to endpoints. Apps using rate limiting should be carefully load tested before deploying. See [Testing endpoints with rate limiting](#test7) in this article for more information.

Reviewer: This has gone into the shared framework now, so I don't think there will be a NuGet package by the time RTM ships: https://github.com/dotnet/aspnetcore/pull/43089/files#diff-bf46b41a392a3ca005eb29dd1c8fcec2f3469729856db52eaf1a4cb0fa910595

## Rate limiter algorithms

The [`RateLimiterOptionsExtensions`](/dotnet/api/microsoft.aspnetcore.ratelimiting.ratelimiteroptionsextensions) class provides the following extension methods for rate limiting:

* [Fixed window](#fixed)
* [Sliding window](#slide)
* [Token bucket](#token)
* [Concurrency](#concur)

<a name="fixed"></a>

### Fixed window limiter

> The [`AddFixedWindowLimiter`](/dotnet/api/microsoft.aspnetcore.ratelimiting.ratelimiteroptionsextensions.addfixedwindowlimiter#microsoft-aspnetcore-ratelimiting-ratelimiteroptionsextensions-addfixedwindowlimiter(microsoft-aspnetcore-ratelimiting-ratelimiteroptions-system-string-system-threading-ratelimiting-fixedwindowratelimiteroptions)) uses a fixed time window to limit requests. When the time window expires, a new time window starts and the request limit is reset.

Reviewer: ```suggestion
The [`AddFixedWindowLimiter`](/dotnet/api/microsoft.aspnetcore.ratelimiting.ratelimiteroptionsextensions.addfixedwindowlimiter#microsoft-aspnetcore-ratelimiting-ratelimiteroptionsextensions-addfixedwindowlimiter(microsoft-aspnetcore-ratelimiting-ratelimiteroptions-system-string-system-threading-ratelimiting-fixedwindowratelimiteroptions)) method uses a fixed time window to limit requests. When the time window expires, a new time window starts and the request limit is reset.
```

Consider the following code:

:::code language="csharp" source="~/../AspNetCore.Docs.Samples/fundamentals/middleware/rate-limit/WebRateLimitAuth/Program.cs" id="snippet_fixed":::

The preceding code:

> * Calls [UseRateLimiter](/dotnet/api/microsoft.aspnetcore.builder.ratelimiterapplicationbuilderextensions.useratelimiter) to enable rate limiting.

Reviewer: This link appears to be broken in the local copy I'm looking at.

![image](https://user-images.githubusercontent.com/1439341/186719351-25dc187c-48b8-47c9-9008-cdc626d16f11.png)

Author: You have to remove `review.` from the URL
https://review.docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.builder.ratelimiterapplicationbuilderextensions.useratelimiter?branch=pr-en-us-26831

* Creates a fixed window limiter with a policy name of `"fixed"` and sets:

> * `permitLimit` to 4 and the time `window` to 12. A maximum of 4 requests per each 12-second window are allowed.

Reviewer 2: Would be easier to read if these bullets were indented one further

Author: @wtgodbe  I don't understand. This is the Microsoft docs standard indenting. I don't see how double indenting would make it easier to read.

<img width="335" alt="image" src="https://user-images.githubusercontent.com/3605364/186993017-841c71d9-ef29-468d-b9b8-8769035c4d4a.png">

* `queueProcessingOrder` to `QueueProcessingOrder.OldestFirst`.
* `queueLimit` to 2.

Apps should use [Configuration](xref:fundamentals/configuration/index) to set limiter options. The preceding code using Configuration:

> :::code language="csharp" source="~/../AspNetCore.Docs.Samples/fundamentals/middleware/rate-limit/WebRateLimitAuth/Program.cs" id="snippet_fixed2":::

Reviewer: Does the code sample for this need to include the code for the `MyRateLimitOptions` object?

<a name="slide"></a>

### Sliding window limiter

> In the sliding window algorithm:

Reviewer: Should this be "A"? _"Is similar..."_ reads a bit weird to be following this sentence as-is otherwise.

```suggestion
A sliding window algorithm:
```

> * Is similar to the fixed window limiter but adds segments per window. The window slides one segment (window)/(segments per window).

Reviewer 2: The end of this line is kinda confusing, maybe be more verbose?

* Limits the requests for a window to `permitLimit` requests.
* Each time window is divided in `n` segments per window.
* Requests taken from the expired time segment one window back (`n` segments prior to the current segment), are added to the current segment. We refer to the most expired time segment one window back as the expired segment.  Consider the following table which shows a sliding window limiter with a 30-second window, 3 segments per window and a limit of 100 requests:

> * The top row and first column shows the time segment.

Reviewer 2: Indentation would help here too

Author: See [Sliding window limiter](https://review.docs.microsoft.com/en-us/aspnet/core/performance/rate-limit?view=aspnetcore-7.0&branch=pr-en-us-26756#sliding-window-limiter)
I don't understand the benefit from deviating from the standard.

* The second row shows the remaining requests available. The remaining requests are available-requests+recycled.
* The third and lower rows show the requests made at that time segment and recycled requests available from the prior expired segment.
* From time 30 on, the request taken from the 3 times slots previous are added back to the request limit.

| Time | 0  | 10  | 20 | 30 | 40 | 50 | 60 |
| ---- | -- | --  | -- | -- | -- | -- | -- |
| Available | 100-20+0=80 | 80-30+0=50  | 50-40+0=10 | 10-30+20=0 |0+30-10=20 | 20-10+40=50 | 50-35+30=45 |
|  0    | -20            |                                  |  |  |  |  | |
|  10   |               | -30                             |  |  |  |  | |
|  20   |               |            | -40                    |  |  |  | |
|  30   | **[+20]**     |            |         | -30        |  |  | |
|  40   |          |**[+30]**|       |                 | -10   |  | |
|  50   |          |           | **[+40]**  |            |               | -10  | |

> |  60   |          |           |            |  **[+30]**  |    |  | -35|

Reviewer 3: I think this table is too confusing

Author: What about something like my updated table? If you don't like that approach, I'll delete it. I also added a couple bullets explaining the blue and red lines.

Reviewer 3: Maybe slightly better, still not a fan though. Maybe let some other folks look at it and see if they understand?

I'm slightly biased, but I like how the algorithm was displayed [here](https://devblogs.microsoft.com/dotnet/announcing-rate-limiting-for-dotnet/#sliding-window-limit).

Author: > Maybe slightly better, still not a fan though. Maybe let some other folks look at it and see if they understand?
> 
> I'm slightly biased, but I like how the algorithm was displayed [here](https://devblogs.microsoft.com/dotnet/announcing-rate-limiting-for-dotnet/#sliding-window-limit).

I had a hard time following that, it struck me as a diagram written by someone very familiar with the algorithm.  At any rate the explanation and diagrams belong in https://github.com/dotnet/docs/issues/30426 which @IEvangelist owns, so he can own that.

IMAO, rate limiting is a gamechanger so I'm going to merge this so it can be live. I'll tweat it  with a challenge to come up with a better diagram.

Author: ![2022-08-30 13 39 11](https://user-images.githubusercontent.com/3605364/187563723-bd99fd9b-7026-4142-9d4f-dda571a903b2.jpg)
@BrennanConroy  I tried the preceding prototype but it looked worse.

| 40   | 90        |  6    | 16    | 100 |
| 50   | 100       | 40    | 20    | 80 |
| 60   | 80        | 50    | 20    | 50 |

The following code uses the token bucket limiter:

:::code language="csharp" source="~/../AspNetCore.Docs.Samples/fundamentals/middleware/rate-limit/WebRateLimitAuth/Program.cs" id="snippet_token":::

When `autoReplenishment` is set to `true`, an internal timer replenishes the tokens every `replenishmentPeriod`; when set to `false`, the app must call `TryReplenish` on the limiter.

<a name="concur"></a>

### Concurrency limiter

The concurrency limiter limits the number concurrent requests. Each request reduces the concurrency limit by one. When a request completes, the limit is increased by one. Unlike the other requests limiters that limit the total number of requests for a specified period, the concurrency limiter limits only the number of concurrent requests and doesn't cap the number of requests in a time period.

The following code uses the concurrency limiter:

:::code language="csharp" source="~/../AspNetCore.Docs.Samples/fundamentals/middleware/rate-limit/WebRateLimitAuth/Program.cs" id="snippet_concur":::

## Limiter algorithm comparison

> The fixed, sliding, and token limiter all limit the maximum number of requests in a time period. The concurrency limiter limits only the number of concurrent requests and doesn't cap the number of requests in a time period. The cost of an endpoint should be considered when selecting a limiter. The cost of an endpoint includes the resources used, for example, data access, CPU, and I/O.

Reviewer: ```suggestion
The fixed, sliding, and token limiters all limit the maximum number of requests in a time period. The concurrency limiter limits only the number of concurrent requests and doesn't cap the number of requests in a time period. The cost of an endpoint should be considered when selecting a limiter. The cost of an endpoint includes the resources used, for example, data access, CPU, and I/O.
```

> The fixed, sliding, and token limiter all limit the maximum number of requests in a time period. The concurrency limiter limits only the number of concurrent requests and doesn't cap the number of requests in a time period. The cost of an endpoint should be considered when selecting a limiter. The cost of an endpoint includes the resources used, for example, data access, CPU, and I/O.

Reviewer: Is time (in the abstract _wibbly-wobbly-timey-wimey_ sense) also worth mentioning as a "cost" consideration?

## Rate limiter samples

> The following samples aren't meant for production code but examples on how to use the limiters.

Reviewer: ```suggestion
The following samples aren't meant for production code but are examples on how to use the limiters.
```

### Limiter with `OnRejected`, `RetryAfter`, and `GlobalLimiter`

The following sample:

> * Creates a [RateLimiterOptions.OnRejected](xref:Microsoft.AspNetCore.RateLimiting.RateLimiterOptions.OnRejected) callback that is called when a request exceeds the specified limit. `retryAfter` can be used with the[`TokenBucketRateLimiter`](https://source.dot.net/#System.Threading.RateLimiting/System/Threading/RateLimiting/TokenBucketRateLimiter.cs), [`FixedWindowLimiter`](https://source.dot.net/#System.Threading.RateLimiting/System/Threading/RateLimiting/FixedWindowRateLimiter.cs), and [`SlidingWindowLimiter`](https://source.dot.net/#System.Threading.RateLimiting/System/Threading/RateLimiting/SlidingWindowRateLimiter.cs) because these algorithms are able to calculate when limits will be added. The `ConcurrencyLimiter` has no way of calculating when permits will be available.

Reviewer 3: ```suggestion
* Creates a [RateLimiterOptions.OnRejected](xref:Microsoft.AspNetCore.RateLimiting.RateLimiterOptions.OnRejected) callback that is called when a request exceeds the specified limit. `retryAfter` can be used with the [`TokenBucketRateLimiter`](https://source.dot.net/#System.Threading.RateLimiting/System/Threading/RateLimiting/TokenBucketRateLimiter.cs), [`FixedWindowLimiter`](https://source.dot.net/#System.Threading.RateLimiting/System/Threading/RateLimiting/FixedWindowRateLimiter.cs), and [`SlidingWindowLimiter`](https://source.dot.net/#System.Threading.RateLimiting/System/Threading/RateLimiting/SlidingWindowRateLimiter.cs) because these algorithms are able to estimate when more permits will be added. The `ConcurrencyLimiter` has no way of calculating when permits will be available.
```

* Adds the following limiters:

  * A `SampleRateLimiterPolicy` which implements the `IRateLimiterPolicy<TPartitionKey>` interface. The `SampleRateLimiterPolicy` class is shown later in this article.
  * A `SlidingWindowLimiter`:
    * With a partition for each authenticated user.

>     * One partition for all anonymous users.

Reviewer 3: ```suggestion
    * One shared partition for all anonymous users.
```

  * A <xref:Microsoft.AspNetCore.RateLimiting.RateLimiterOptions.GlobalLimiter> that is applied to all requests. The global limiter will be executed first, followed by the endpoint-specific limiter, if one exists. The `GlobalLimiter` creates a partition for each <xref:System.Net.IPAddress>.

:::code language="csharp" source="~/../AspNetCore.Docs.Samples/fundamentals/middleware/rate-limit/WebRateLimitAuth/Program.cs" id="snippet":::

> > [!WARNING]

Reviewer: There's a random `2` is the local version I have when I view this:

![image](https://user-images.githubusercontent.com/1439341/186720842-0b7d55a9-d0f5-48c3-a2bf-30168f487c95.png)

Author: I fixed that and didn't update your copy. It's updated now with my latest changes. Thanks for your comments and suggestions, they were very helpful.

>Creating partitions on client IP addresses makes the app vulnerable to Denial of Service Attacks which employ IP Source Address Spoofing. For more information, see [BCP 38 RFC 2827 Network Ingress Filtering: Defeating Denial of Service Attacks which employ IP Source Address Spoofing](https://www.rfc-editor.org/info/bcp38).

See [the samples repository for the complete `Program.cs`](https://github.com/dotnet/AspNetCore.Docs.Samples/blob/main/fundamentals/middleware/rate-limit/WebRateLimitAuth/Program.cs#L145,L281) file.

The `SampleRateLimiterPolicy` class

:::code language="csharp" source="~/../AspNetCore.Docs.Samples/fundamentals/middleware/rate-limit/WebRateLimitAuth/SampleRateLimiterPolicy.cs" id="snippet_1":::

In the preceding code, <xref:Microsoft.AspNetCore.RateLimiting.RateLimiterOptions.OnRejected> uses <xref:Microsoft.AspNetCore.RateLimiting.OnRejectedContext> to set the response status to [429 Too Many Requests](https://developer.mozilla.org/docs/Web/HTTP/Status/429). The default rejected status is [503 Service Unavailable](https://developer.mozilla.org/docs/Web/HTTP/Status/503).

> :::code language="csharp" source="~/../AspNetCore.Docs.Samples/fundamentals/middleware/rate-limit/WebRateLimitAuth/SampleRateLimiterPolicy.cs" id="snippet":::

Reviewer 3: This is just showing a slightly truncated version of the previous snippet, not sure it's adding much.

<a name="test7"></a>

## Testing endpoints with rate limiting

Before deploying an app using rate limiting to production, it's a good idea to stress test the app to validate the rate limiters and options used. For example, create a [JMeter script](https://jmeter.apache.org/usermanual/jmeter_proxy_step_by_step.html) with a tool like [BlazeMeter](https://guide.blazemeter.com/hc/articles/207421695-Writing-your-first-JMeter-script) or [Apache JMeter HTTP(S) Test Script Recorder](https://jmeter.apache.org/usermanual/jmeter_proxy_step_by_step.html) and load the script to [Azure Load Testing](/azure/load-testing/overview-what-is-azure-load-testing).

Creating partitions with user input makes the app vulnerable to [Denial of Service](https://www.cisa.gov/uscert/ncas/tips/ST04-015) (DoS) Attacks. For example, creating partitions on client IP addresses makes the app vulnerable to Denial of Service Attacks that employ IP Source Address Spoofing. For more information, see [BCP 38 RFC 2827 Network Ingress Filtering: Defeating Denial of Service Attacks that employ IP Source Address Spoofing](https://www.rfc-editor.org/info/bcp38).

> :::moniker-end

Reviewer: There's no mention of the No-Op policy in this document, assuming it's still going to be shipped. If it is, worth also documenting it?

Author: Per [this comment](https://github.com/dotnet/AspNetCore.Docs.Samples/pull/13#discussion_r949659464)

> Adding a no limiter is functionally equivalent to not adding any endpoint limiter for this endpoint - the global limiter will still run. We're removing `in rc1 (we are also adding an attribute that allows you to disable rate limiting entirely for an endpoint, including the global limiter)

---
title: Rate limiting middleware in ASP.NET Core
author: rick-anderson
description: Learn how limit requests in ASP.NET Core apps
ms.author: riande
ms.custom: mvc
ms.date: 4/05/2019
uid: performance/rate-limit
---

# Rate limiting middleware in ASP.NET Core

> By [Arvin Kahbazi](https://github.com/Kahbazi) and [Rick Anderson](https://twitter.com/RickAndMSFT)

Author: @Kahbazi I'm just starting on this but I'll be asking you questions soon.
